pageserver/client_grpc/src/pool.rs

```diff
@@ -407,287 +422,187 @@ impl Drop for ClientGuard {
+    stream: Option<BiStream>,             // Some until dropped
+    active: bool,                         // not returned to pool if true
     permit: Option<OwnedSemaphorePermit>, // None if pool is unbounded
 }
 
 impl StreamGuard {
-    /// Sends a request on the stream and awaits the response. Consumes the guard, since it's only
-    /// valid for a single request (to enforce queue depth). This also drops the guard on return and
-    /// returns the queue depth quota to the pool.
+    /// Sends a request on the stream and awaits the response. If the future is dropped before it
+    /// resolves (e.g. due to a timeout or cancellation), the stream will be closed to cancel the
+    /// request and is not returned to the pool. The same is true if the stream errors, in which
+    /// case the caller can't send further requests on the stream.
     ///
-    /// The `GetPageRequest::request_id` must be unique across in-flight requests.
+    /// We only support sending a single request at a time, to eliminate head-of-line blocking. See
+    /// module documentation for details.
     ///
     /// NB: errors are often returned as `GetPageResponse::status_code` instead of `tonic::Status`
     /// to avoid tearing down the stream for per-request errors. Callers must check this.
     pub async fn send(
-        self,
+        &mut self,
         req: page_api::GetPageRequest,
     ) -> tonic::Result<page_api::GetPageResponse> {
-        let (resp_tx, resp_rx) = oneshot::channel();
+        let req_id = req.request_id;
+        let stream = self.stream.as_mut().expect("not dropped");
 
-        self.sender
-            .send((req, resp_tx))
-            .await
+        // Mark the stream as active. We only allow one request at a time, to avoid HoL-blocking.
+        // We also don't allow reuse of the stream after an error.
+        //
+        // NB: this uses a watch channel, so it's unsafe to change this code to pipeline requests.
+        if self.active {
+            return Err(tonic::Status::internal("stream already active"));
+        }
+        self.active = true;
+
+        // Send the request and receive the response. If the stream errors for whatever reason, we
+        // don't reset `active` such that the stream won't be returned to the pool.
+        stream
+            .sender
+            .send(req)
             .map_err(|_| tonic::Status::unavailable("stream closed"))?;
 
-        resp_rx
+        let resp = stream
+            .receiver
+            .next()
             .await
-            .map_err(|_| tonic::Status::unavailable("stream closed"))?
+            .ok_or_else(|| tonic::Status::unavailable("stream closed"))??;
+
+        if resp.request_id != req_id {
+            return Err(tonic::Status::internal(format!(
+                "response ID {} does not match request ID {}",
+                resp.request_id, req_id
+            )));
+        }
+
+        // Success, mark the stream as inactive.
+        self.active = false;
```

Comment:
nit: not sure active is the best name. Technically it's not doing anything if the stream failed. Maybe `return_to_pool` or `eligible_for_reuse` would be clearer.

Reply:
Fair, changed to `can_reuse`.